taxonomy/utils.py

```diff
@@ -601,3 +602,61 @@ def apply_batching_to_translate_large_text(key, source_text):
     result['TranslatedText'] = translated_text
     result['SourceLanguageCode'] = source_language_code
     return result
+
+
+def duplicate_model_instance(source_block):
+    """
+    Duplicate passed django model as described in django docs.
+
+    https://docs.djangoproject.com/en/4.1/topics/db/queries/#copying-model-instances
+    """
+    source_block.id = None
+    source_block.pk = None
+    source_block._state.adding = True  # pylint: disable=protected-access
+    if hasattr(source_block, "created"):
+        source_block.created = now()
+    if hasattr(source_block, "modified"):
+        source_block.modified = now()
+    source_block.save()
+    return source_block
+
+
+def delete_product(key_or_uuid: str, product_type: ProductTypes):
+    """
+    Delete product from database if it exists.
+
+        key_or_uuid (str): Key or uuid of the product to be deleted.
+        product_type (ProductTypes): Product type.
+    """
+    product_model, identifier = get_product_skill_model_and_identifier(product_type)
+    product_model.objects.filter(**{identifier: key_or_uuid}).delete()
+
+
+def duplicate_xblock_skills(source_xblock_uuid, xblock_uuid):
+    """
+    Duplicate xblock and its skills if source xblock exists.
+
+        source_xblock_uuid (str): source xblock usage key.
+        xblock_uuid (str): new xblock usage key.
+    """
+    # get source xblock_skill instance.
+    source_xblock = XBlockSkills.objects.filter(usage_key=source_xblock_uuid).first()
+    if not source_xblock:
+        LOGGER.info(f'[TAXONOMY] Source xblock: {source_xblock_uuid} not found')
+        return
+
+    # just in case xblock_skill with new usage_key exists, stop execution.
+    if XBlockSkills.objects.filter(usage_key=xblock_uuid).exists():
```

Comment:
When would this actually happen? I can see a race condition in case of some "update" function. When would a race condition arise for a "duplicate" action?

Also, in case it occurs - should we update the existing model with latest values from the original block or just not do anything and persist the old values?

Reply:
It would be better to avoid adding code for hypothetical scenarios. However, I think this can happen when the signal that triggers this is received twice. While it might be rare, it seems like it can occur in [Kafka and there is no way to avoid it](https://stackoverflow.com/questions/29647656/effective-strategy-to-avoid-duplicate-messages-in-apache-kafka-consumer). So, let us leave it for now.